File: JavaPartI/la_maxProductOfCuttingRope.java
/// dp - dp[i]: max product of rope with length of i, with at least one cut
// sol1:  left big seg, right big seg
public class Solution {
  public int maxProduct(int length) {
    // Write your solution here.
    if (length == 0) {
      return 0;
    }
    int n = length;
    int[] dp = new int[n + 1];
    dp[1] = 1;
    for (int i = 2; i <= n; i++) {
      for (int k = 1; k <= i / 2; k++) {
        dp[i] = Math.max(dp[i], Math.max(dp[k], k) * Math.max(dp[i - k], i - k));
      }
    }
    return dp[n];
  }
}
/// sol2 : dp , left bit seg, right small seg



// recursion + left big, right small

File: pythonPart/la_maxProductOfCuttingRope.py
class Solution(object):
  def maxProduct(self, length):
    """
    input: int length
    return: int
    """
    # write your solution here

    n = length
    M = [0] * (n + 1)
    M[1] = 1
    for i in range(2, n + 1):  # total length of each case
      for l in range(1, i // 2 + 1):
        r = i - l
        M[i] = max(M[i], max(l, M[l]) * max(r, M[r]))
    return M[n]
